addons/l10n_in/models/account_invoice.py

```diff
@@ -71,18 +71,29 @@ def _compute_l10n_in_state_id(self):
 
     @api.depends('l10n_in_state_id')
     def _compute_fiscal_position_id(self):
-        other_country = None
-        in_moves = self.filtered(lambda move: move.country_code == 'IN' and move.is_sale_document(include_receipts=True))
+        foreign_state = None
+        in_moves = self.filtered(lambda move: move.country_code == 'IN' and move.is_invoice(include_receipts=True))
         for move in in_moves:
-            state_id = move.l10n_in_state_id.id
-            country_id = move.l10n_in_state_id.country_id.id
-            if move.l10n_in_state_id.l10n_in_tin == '96':  # Foreign Country
-                other_country = other_country or self.env['res.country'].search([('code', '!=', 'IN')], limit=1)
-                country_id = other_country.id
-                state_id = False
+            state_id = move.l10n_in_state_id
+            if move.is_purchase_document(include_receipts=True):
+                if state_id == move.partner_id.state_id:  # Within State
+                    state_id = move.company_id.state_id
+                else:  # Inter State or Export/SEZ
+                    operator = '=' if state_id.l10n_in_tin != '96' and move.partner_id.country_id.code == 'IN' else '!='
+                    state_id = self.env['res.country.state'].search([
+                        ('country_id', operator, self.env.ref('base.in').id),
+                        ('id', '!=', move.company_id.state_id.id)
+                    ], limit=1)
```

Comment:
```suggestion
                elif state_id == move.company_id.state_id:
                    # To apply inter state tax for purchase if place of supply is same as company state
                    state_id = self.env['res.country.state'].search([
                        ('country_id', '=', self.env.ref('base.in').id),
                        ('id', '!=', move.company_id.state_id.id)
                    ], limit=1)
```
I don't see the point to handle `OC` differently for sale and purchase?